srai/joiners/intersection_joiner.py

```diff
@@ -0,0 +1,58 @@
+"""
+Intersection Joiner.
+
+This module contains intersection joiner implementation.
+
+"""
+
+import geopandas as gpd
+import pandas as pd
+
+from srai.joiners import JoinerResult
+
+
+class IntersectionJoiner:
+    """
+    Intersection Joiner.
+
+    Intersection Joiner allows to join two GeoDataFrames and find all
+    overlapping geometries. It does not apply any grouping or
+    aggregation.
+
+    """
+
+    def join(self, regions: gpd.GeoDataFrame, features: gpd.GeoDataFrame) -> JoinerResult:
+        """
+        Join features to regions.
+
+        Args:
+            regions (gpd.GeoDataFrame): regions with which features are joined
+            features (gpd.GeoDataFrame): features to be joined
+
+        Returns:
+            JoinerResult object, which contains 3 GeoDataFrames:
+            * original regions
+            * original features
+            * intersection of regions and features, which contains a MultiIndex and a geometry with
+              the intersection
+
+        """
+        joined_parts = []
+
+        for _, single in features.groupby(features["geometry"].geom_type):
+            joined_parts.append(
+                gpd.overlay(
+                    single[["geometry"]].reset_index().rename(columns={"index": "features_id"}),
+                    regions[["geometry"]].reset_index().rename(columns={"index": "region_id"}),
```

Comment:
Here I assume that both gdf's have unnamed index with correct values (as we decided to use index). Should I remove the name if provided (but this will change input gdf) or just assume that gdf is correct?

Reply:
I'd make it as robust as possible. Best option would be to rename the index if it intervenes with this operation like this: `df.reset_index().rename(columns={df.index.name:'bar'})`.
https://stackoverflow.com/a/40914267/7766101

Reply:
The problem is, that when you apply `.reset_index()` when there is no name, the new column will be named `"index"`, otherwise `df.index.name`. And you cannot tell if something went wrong, because `.rename()` will not tell that it could not find a particular column.

An option is to remove name before, but this may either alter original gdf of increase memory consumption. I could also add second element to this mapping (`.rename(columns={"index": "bar", df.index.name: "bar"})`, but this may be problematic in some edge cases, where there is a column named "index"

Reply:
You can rename it on the fly according to the documentation:

```python
single[["geometry"]].reset_index(names="features_id")
regions[["geometry"]].reset_index(names="region_id")
```

It will fail if there is already a column named like that or if the index is multileveled.

Reply:
Additionally, `features` probably should be singularized into `feature_id`.